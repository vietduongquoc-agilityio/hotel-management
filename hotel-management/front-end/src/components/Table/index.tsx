import { Box, Text, UnorderedList, ListItem, Alert } from "@chakra-ui/react";
import { memo, useEffect, useState } from "react";

// Interfaces
import {
  DeleteFunctionType,
  EditFunctionType,
  RateData,
  RoomData,
  DealData,
} from "@/interfaces";

// Components
import {
  Button,
  DeleteRate,
  DeleteRoom,
  EditRateModal,
  EditRoomModal,
} from "@/components";

// utils
import {
  tableHeaders,
  renderRoomBody,
  renderRateBody,
  renderDealBody,
} from "@/utils";

export interface TableProps<T> {
  data: T[];
  type: "room" | "rate" | "guest" | "deal";
  error?: string | null;
  onDelete: (id: string) => void;
  onEdit: (updatedData: T) => void;
}

const Table = <T extends RoomData | RateData | DealData>({
  error,
  data,
  type,
  onDelete,
  onEdit,
}: TableProps<T>) => {
  const [activeId, setActiveId] = useState<string | null>(null);
  const editFunction = onEdit as <T>(updatedData: T) => void;

  useEffect(() => {
    const handleClickOutside = (event: MouseEvent) => {
      if (activeId) {
        const menuElement = document.getElementById(`${activeId}`);
        if (menuElement && !menuElement.contains(event.target as Node)) {
          setActiveId(null);
        }
      }
    };
    window.addEventListener("click", handleClickOutside);
    return () => {
      window.removeEventListener("click", handleClickOutside);
    };
  }, [activeId]);

  const toggleMenu = (id: string) => {
    setActiveId((prev) => (prev === id ? null : id));
  };

  if (error) return <Alert status="error">{error}</Alert>;
  if (data.length === 0)
    return (
      <Alert
        fontWeight="500"
        color="grey.800"
        borderRadius="10px"
        status="info"
      >
        No {type}s available.
      </Alert>
    );

  const renderRow = (item: T) => {
    const result =
      type === "room"
        ? renderRoomBody(item as RoomData)
        : type === "rate"
        ? renderRateBody(item as RateData)
        : renderDealBody(item as DealData);

    return (
      <>
        {result.map((item, index) => (
          <Text
            key={index + `${type}-body`}
            w={item.width}
            color={item.color}
            fontWeight={item.fontWeight}
            borderRadius={item.borderRadius}
            p={item.p}
            m={item?.m}
            bg={item.bg}
            display={item.display}
            justifyContent={item.justifyContent}
            fontSize={item?.fontSize}
          >
            {item.value instanceof Date
              ? item.value.toLocaleString()
              : item.value}
          </Text>
        ))}
      </>
    );
  };

  const toggleStyle = type === "room" ? "auto" : "20%";

  const EditFunctionModal = ({
    initialData,
    onEditFunction,
  }: EditFunctionType<T>) => {
    return type === "room" ? (
      <EditRoomModal
        initialRoomData={initialData as RoomData}
        onEditRoom={onEditFunction}
      />
    ) : (
      <EditRateModal
        initialRateData={initialData as RateData}
        onEditRate={onEditFunction}
      />
    );
  };

  const DeleteFunctionModal = ({
    documentId,
    onDeleteFunction,
  }: DeleteFunctionType) => {
    return type === "room" ? (
      <DeleteRoom roomId={documentId} onDeleteRoom={onDeleteFunction} />
    ) : (
      <DeleteRate rateId={documentId} onDeleteRate={onDeleteFunction} />
    );
  };

  return (
    <Box
      borderTopLeftRadius="8px"
      borderTopRightRadius="8px"
      border="1px solid #d4e5fa"
    >
      <UnorderedList
        display="flex"
        maxW="1020px"
        w="100%"
        m="0"
        bg="grey.50"
        fontSize="12px"
        fontWeight="500"
        color="grey.500"
        borderTopLeftRadius="8px"
        borderTopRightRadius="8px"
        p="10px 24px"
      >
        {tableHeaders(type).map((header, index) => (
          <ListItem
            key={index}
            w={header.width}
            listStyleType="none"
            ml={header.ml}
          >
            {header.label}
          </ListItem>
        ))}
      </UnorderedList>
      {data.map((item: T) => {
        return (
          <Box
            key={item.documentId}
            fontSize="14px"
            fontWeight="400"
            display="flex"
            maxW="1020px"
            w="100%"
            p="17px 24px"
            position="relative"
            border="1px solid #d4e5fa"
          >
            {renderRow(item)}

            <Box display="flex" justifyContent="flex-end" w={toggleStyle}>
              <Button
                bg="white.200"
                color="grey.800"
                _hover={{ bg: "white.200" }}
                height="15px"
                onClick={(e) => {
                  e.stopPropagation();
                  toggleMenu(item.documentId);
                }}
                text={"â‹®"}
                buttonType={"primary"}
                variant={"contained"}
                colorScheme={"primary"}
              />
            </Box>

            {activeId === item.documentId && (
              <Box
                id={item.documentId}
                top="25px"
                right="70px"
                position="absolute"
                backgroundColor="white.200"
                border="1px solid #989fad"
                p="7px"
                boxShadow="0px 4px 8px rgba(57, 56, 56, 0.466)"
                display="flex"
                flexDirection="column"
                gap="10px"
                zIndex="100"
                borderRadius="8px"
                w="80px"
              >
                <EditFunctionModal
                  initialData={item}
                  onEditFunction={editFunction}
                />
                <DeleteFunctionModal
                  documentId={item.documentId}
                  onDeleteFunction={onDelete}
                />
              </Box>
            )}
          </Box>
        );
      })}
    </Box>
  );
};

const genericMemo: <T>(component: T) => T = memo;

export default genericMemo(Table);
