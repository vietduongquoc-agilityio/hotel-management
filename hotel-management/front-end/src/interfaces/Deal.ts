export interface DealData {
  dealName: string;
  referenceNumber: string;
  startDate: Date;
  endDate: Date;
  roomType: string;
  statusDeal: string;
  reservationsLeft: number;
  documentId: string;
}

export type NewDealData = Omit<DealData, "documentId">;
