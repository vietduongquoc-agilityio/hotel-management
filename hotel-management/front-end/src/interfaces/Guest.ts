export interface GuestData {
    guestName: string;
    roomType: string;
    stay: number;
    price: number;
    registrationNumber: string;
    totalAmount: number;
    checkInDate: Date;
    documentId: string;
}

export type NewGuestData = Omit<GuestData, "documentId">;
