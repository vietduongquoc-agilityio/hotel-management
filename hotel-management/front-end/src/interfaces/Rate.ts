export interface RateData {
  roomType: string;
  cancellationPolicy: string;
  deals: string;
  dealPrice: string;
  availability: string;
  documentId: string;
  rate: string;
  totalOfBooked?: number;
}

export type NewRateData = Omit<RateData, "documentId">;
