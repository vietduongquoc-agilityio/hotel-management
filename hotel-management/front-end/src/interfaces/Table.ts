export interface EditFunctionType<T> {
  initialData: T;
  onEditFunction: (updatedData: T) => void;
}

export interface DeleteFunctionType {
  documentId: string;
  onDeleteFunction: (id: string) => void;
}
