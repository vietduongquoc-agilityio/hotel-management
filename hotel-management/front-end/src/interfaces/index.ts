export * from "./Rate";
export * from "./Room";
export * from "./Table";
export * from "./Guest"
