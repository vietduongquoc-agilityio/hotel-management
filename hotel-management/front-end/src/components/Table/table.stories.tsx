import { Meta, StoryFn } from "@storybook/react";
import { ChakraProvider } from "@chakra-ui/react";
import { MemoryRouter } from "react-router-dom";
import { themeColor } from "@/themes/Base/colors";

// Components
import Table, { TableProps } from ".";

// InterFace
import { RoomData, RateData, GuestData, DealData } from "@/interfaces";

// Mock Data
const roomData: RoomData[] = [
  {
    documentId: "1",
    roomNumber: "101",
    bedType: "Queen",
    roomFloor: "1",
    roomFacility: "TV, WiFi, Air Conditioner",
    roomStatus: "Available",
  },
  {
    documentId: "2",
    roomNumber: "102",
    bedType: "King",
    roomFloor: "2",
    roomFacility: "TV, WiFi",
    roomStatus: "Booked",
  },
];

const rateData: RateData[] = [
  {
    documentId: "1",
    roomType: "Deluxe",
    deals: "Family Deal",
    cancellationPolicy: "Flexible",
    dealPrice: "120",
    rate: "150",
    totalOfRooms: 10,
    totalOfBooked: 5,
  },
  {
    documentId: "2",
    roomType: "Standard",
    deals: "No Deals",
    cancellationPolicy: "Non-refundable",
    dealPrice: "80",
    rate: "100",
    totalOfRooms: 5,
    totalOfBooked: 5,
  },
];

const guestData: GuestData[] = [
  {
    roomType: "King",
    guestName: "Alex",
    stay: 2,
    price: 1000,
    registrationNumber: "#2222",
    totalAmount: 2000,
    checkInDate: new Date("2024-12-01"),
    documentId: "1",
  },
  {
    roomType: "Queen",
    guestName: "Anna",
    stay: 4,
    price: 1000,
    registrationNumber: "#1111",
    totalAmount: 4000,
    checkInDate: new Date("2024-12-05"),
    documentId: "2",
  },
];

const dealData: DealData[] = [
  {
    dealName: "Alex",
    referenceNumber: "1111",
    startDate: new Date("2024-12-24"),
    endDate: new Date("2024-12-26"),
    roomType: "King",
    statusDeal: "Full",
    reservationsLeft: 2,
    documentId: "1"
  },
  {
    dealName: "Anna",
    referenceNumber: "2222",
    startDate: new Date("2024-12-27"),
    endDate: new Date("2024-12-29"),
    roomType: "Queen",
    statusDeal: "Ongoing",
    reservationsLeft: 2,
    documentId: "2"
  },
];

// Handlers
const handleDelete = (id: string) => alert(`Deleted item with ID: ${id}`);
const handleEdit = (updatedData: RoomData | RateData | GuestData | DealData) =>
  alert(`Edited item: ${JSON.stringify(updatedData, null, 2)}`);

// Meta
export default {
  title: "Components/Table",
  component: Table,
  argTypes: {
    type: {
      control: { type: "select", options: ["room", "rate", "deal", "guest"] },
    },
  },
  tags: ["autodocs"],
  decorators: [
    (Story) => (
      <ChakraProvider theme={themeColor}>
        <MemoryRouter>{Story()}</MemoryRouter>
      </ChakraProvider>
    ),
  ],
} as Meta;

// Template
const Template: StoryFn<TableProps<RoomData | RateData | GuestData | DealData>> = (
  args
) => <Table {...args} />;

// Stories
export const RoomTable = Template.bind({});
RoomTable.args = {
  data: roomData,
  type: "room",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const RateTable = Template.bind({});
RateTable.args = {
  data: rateData,
  type: "rate",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const EmptyRoomTable = Template.bind({});
EmptyRoomTable.args = {
  data: [],
  type: "room",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const EmptyRateTable = Template.bind({});
EmptyRateTable.args = {
  data: [],
  type: "rate",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const ErrorState = Template.bind({});
ErrorState.args = {
  data: [],
  type: "room",
  error: "Failed to fetch data. Please try again later.",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const GuestTable = Template.bind({});
GuestTable.args = {
  data: guestData,
  type: "guest",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const EmptyGuestTable = Template.bind({});
EmptyGuestTable.args = {
  data: [],
  type: "guest",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const DealTable = Template.bind({});
DealTable.args = {
  data: dealData,
  type: "deal",
  onDelete: handleDelete,
  onEdit: handleEdit,
};

export const EmptyDealTable = Template.bind({});
EmptyDealTable.args = {
  data: [],
  type: "deal",
  onDelete: handleDelete,
  onEdit: handleEdit,
};
