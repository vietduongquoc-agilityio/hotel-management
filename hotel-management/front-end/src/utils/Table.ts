// Constant
import { roomStatusBackgrounds, roomStatusColors } from "@/constant";

// Interfaces
import { RoomData, RateData } from "@/interfaces";

export const renderRoomBody = (room: RoomData) => {
  const { roomNumber, bedType, roomFloor, roomFacility, roomStatus } = room;
  return [
    {
      value: roomNumber,
      width: "15%",
      color: "grey.900",
      fontWeight: "500",
    },
    {
      value: bedType + " Bed",
      width: "20%",
    },
    {
      value: roomFloor + " Floor",
      width: "15%",
    },
    {
      value: roomFacility,
      width: "27%",
    },
    {
      value: roomStatus,
      width: "65px",
      borderRadius: "16px",
      p: "4px 8px",
      m: "auto",
      bg: roomStatusBackgrounds[roomStatus],
      display: "flex",
      justifyContent: "center",
      color: roomStatusColors[roomStatus],
      fontSize: "12px",
    },
  ];
};

export const renderRateBody = (rate: RateData) => {
  const {
    roomType,
    deals,
    cancellationPolicy,
    dealPrice,
    totalOfRooms,
    totalOfBooked,
  } = rate;

  const availability = totalOfRooms - totalOfBooked;
  const isFull = availability === 0;
  const textColor = isFull ? "red.500" : "blue.500";
  const backgroundColor = isFull ? "red.100" : "blue.100";

  return [
    {
      value: roomType,
      width: "15%",
      color: "grey.900",
    },
    {
      value: deals,
      width: "15%",
    },
    {
      value: cancellationPolicy,
      width: "15%",
    },
    {
      value: dealPrice,
      width: "15%",
    },
    {
      value: dealPrice,
      width: "15%",
      color: "grey.900",
      fontWeight: "500",
    },
    {
      value: isFull ? "Full" : availability,
      width: "65px",
      color: textColor,
      bg: backgroundColor,
      display: "flex",
      justifyContent: "center",
      p: "2px 4px",
      borderRadius: "16px",
      fontSize: "",
      m: "",
    },
  ];
};

export const tableHeaders = (type: string) => {
  return type === "room"
    ? [
        { label: "Room number", width: "15%" },
        { label: "Bed type", width: "20%" },
        { label: "Room floor", width: "15%" },
        { label: "Room facility", width: "32%" },
        { label: "Status", width: "16%", ml: "12px" },
      ]
    : [
        { label: "Room type", width: "15%" },
        { label: "Deals", width: "15%" },
        { label: "Cancellation policy", width: "15%" },
        { label: "Deal price", width: "15%" },
        { label: "Rate", width: "15%" },
        { label: "Availability", width: "26%" },
      ];
};
