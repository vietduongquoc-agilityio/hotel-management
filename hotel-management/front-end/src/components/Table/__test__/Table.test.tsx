import { render, screen } from "@testing-library/react";
import Table from "../index";
import { RateData, RoomData } from "@/interfaces";

describe("Table Component", () => {
  const mockOnDelete = jest.fn();
  const mockOnEdit = jest.fn();

  const roomData: RoomData[] = [
    {
      documentId: "1",
      roomNumber: "101",
      bedType: "Single",
      roomFloor: "1",
      roomFacility: "Air Conditioning",
      roomStatus: "Available",
    },
  ];

  const rateData: RateData[] = [
    {
      documentId: "1",
      roomType: "Deluxe",
      deals: "Summer Deal",
      cancellationPolicy: "Free Cancellation",
      dealPrice: "200",
      rate: "200",
      totalOfRooms: 10,
      totalOfBooked: 5,
    },
  ];

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders table headers for rooms", () => {
    const { container } = render(
      <Table
        data={roomData}
        type="room"
        onDelete={mockOnDelete}
        onEdit={mockOnEdit}
      />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Room number")).toBeInTheDocument();
    expect(screen.getByText("Bed type")).toBeInTheDocument();
    expect(screen.getByText("Room floor")).toBeInTheDocument();
    expect(screen.getByText("Room facility")).toBeInTheDocument();
    expect(screen.getByText("Status")).toBeInTheDocument();
  });

  it("renders table headers for rates", () => {
    const { container } = render(
      <Table
        data={rateData}
        type="rate"
        onDelete={mockOnDelete}
        onEdit={mockOnEdit}
      />
    );
    expect(container).toMatchSnapshot();
    expect(screen.getByText("Room type")).toBeInTheDocument();
    expect(screen.getByText("Deals")).toBeInTheDocument();
    expect(screen.getByText("Cancellation policy")).toBeInTheDocument();
    expect(screen.getByText("Deal price")).toBeInTheDocument();
    expect(screen.getByText("Rate")).toBeInTheDocument();
    expect(screen.getByText("Availability")).toBeInTheDocument();
  });

  it("displays an error message when error is provided", () => {
    render(
      <Table
        data={[]}
        type="room"
        error="Failed to fetch data"
        onDelete={mockOnDelete}
        onEdit={mockOnEdit}
      />
    );

    expect(screen.getByText("Failed to fetch data")).toBeInTheDocument();
  });

  it("displays no items message when data is empty", () => {
    render(
      <Table
        data={[]}
        type="rate"
        onDelete={mockOnDelete}
        onEdit={mockOnEdit}
      />
    );

    expect(screen.getByText("No rates available.")).toBeInTheDocument();
  });
});
