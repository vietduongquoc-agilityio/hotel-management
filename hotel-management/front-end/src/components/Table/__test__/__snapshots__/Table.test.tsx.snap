// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table Component renders table headers for rates 1`] = `
<div>
  <div
    class="css-1y7mht"
  >
    <ul
      class="css-1w40mxc"
      role="list"
    >
      <li
        class="css-10iu03d"
      >
        Room type
      </li>
      <li
        class="css-10iu03d"
      >
        Deals
      </li>
      <li
        class="css-10iu03d"
      >
        Cancellation policy
      </li>
      <li
        class="css-10iu03d"
      >
        Deal price
      </li>
      <li
        class="css-10iu03d"
      >
        Rate
      </li>
      <li
        class="css-18knwf5"
      >
        Availability
      </li>
    </ul>
    <div
      class="css-m3cdxu"
    >
      <p
        class="chakra-text css-1oujbce"
      >
        Deluxe
      </p>
      <p
        class="chakra-text css-w56oqh"
      >
        Summer Deal
      </p>
      <p
        class="chakra-text css-w56oqh"
      >
        Free Cancellation
      </p>
      <p
        class="chakra-text css-w56oqh"
      >
        200
      </p>
      <p
        class="chakra-text css-1dmpze8"
      >
        200
      </p>
      <p
        class="chakra-text css-27rlwr"
      >
        5
      </p>
      <div
        class="css-u8uklq"
      >
        <button
          class="chakra-button css-1d4fhah"
          type="button"
        >
          ⋮
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Table Component renders table headers for rooms 1`] = `
<div>
  <div
    class="css-1y7mht"
  >
    <ul
      class="css-1w40mxc"
      role="list"
    >
      <li
        class="css-10iu03d"
      >
        Room number
      </li>
      <li
        class="css-17i31e1"
      >
        Bed type
      </li>
      <li
        class="css-10iu03d"
      >
        Room floor
      </li>
      <li
        class="css-qn161l"
      >
        Room facility
      </li>
      <li
        class="css-14uvbdo"
      >
        Status
      </li>
    </ul>
    <div
      class="css-m3cdxu"
    >
      <p
        class="chakra-text css-1dmpze8"
      >
        101
      </p>
      <p
        class="chakra-text css-ufscxh"
      >
        Single Bed
      </p>
      <p
        class="chakra-text css-w56oqh"
      >
        1 Floor
      </p>
      <p
        class="chakra-text css-v8ppkx"
      >
        Air Conditioning
      </p>
      <p
        class="chakra-text css-1hcog7o"
      >
        Available
      </p>
      <div
        class="css-ekszn"
      >
        <button
          class="chakra-button css-1d4fhah"
          type="button"
        >
          ⋮
        </button>
      </div>
    </div>
  </div>
</div>
`;
